### DISCUSSION TITLE
rust_analyzer failed to discover workspace LSP

### DISCUSSION THREAD

**Discussion by JamesSayegh** [LazyVim/LazyVim#261] [Q&A] [answered]
Hello! 

i recently installed rust_analyzer for my rust projects but it keeps giving me the same failed to discover workspace error message.

yes i did initialize my project through cargo new, this issue is happening in every rust project i try to work on 

any ideas on what i could do?

UPDATE:

looking at LspInfo it shows that the root directory only goes into /home/$USER, doing a cargo new in my home directory gets rust analyzer to work, how can i change the root directory to be in any folder i cd into?

**Comment by folke** [comment]
That's not how lsp works. Check the lspconfig docs

**Comment by dpetka2001** [comment]
I installed `rust-analyzer` from `mason` and didn't need to do any further configuration and it recognizes the `roor_dir` just fine whenever i `cd` into a directory and `cargo new` to start a new project. The `root_dir` gets discovered by corresponding `Cargo.toml` each time which is the pattern `rust-analyzer` uses to discover the workspace.

**Reply by folke** [reply]
I think op wants the root_dir to be his cwd, without doing `cargo new`. Probably doesn't work like that....

**Reply by dpetka2001** [reply]
He could try adding manually a `Cargo.toml` in his cwd if that's what he wants. Either way `Cargo.toml` must exist at the root of his project for `rust-analyzer` to be able to discover the workspace. But I don't know how "good" practice would that be considered(adding `Cargo.toml` manually)

**Comment by JamesSayegh** [comment, marked as answer]
> Hello!
> 
> i recently installed rust_analyzer for my rust projects but it keeps giving me the same failed to discover workspace error message.
> 
> yes i did initialize my project through cargo new, this issue is happening in every rust project i try to work on
> 
> any ideas on what i could do?
> 
> UPDATE:
> 
> looking at LspInfo it shows that the root directory only goes into /home/$USER, doing a cargo new in my home directory gets rust analyzer to work, how can i change the root directory to be in any folder i cd into?

Apologies for the confusion, i tried creating a rust project directly in my home directory and now for some reason rust_analyzer simply just works everywhere again? really weird behavior but im glad it works now

**Comment by dpetka2001** [comment]
Did you delete the project from your home directory? If not maybe it sees the `Cargo.toml` there and discovers your whole home directory as the workspace. However, it shouldn't be like that. It should be in every root directory of every project created with `cargo new`. If you deleted the project from your home directory and now works in every other directory regardless then glad for your good luck getting it to work :smiley:
